controller/openebs/cstor.go

```diff
@@ -111,6 +118,354 @@ func (p *Planner) setCStorDefaultsIfNotSet() error {
+				p.ObservedOpenEBS.Spec.CstorConfig.CSI.CSINode.ISCSIPath, "spec", "mountPath")
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// update the containers
+	updateContainer := func(obj *unstructured.Unstructured) error {
+		containerName, _, err := unstructured.NestedString(obj.Object, "spec", "name")
+		if err != nil {
+			return err
+		}
+		envs, _, err := unstruct.GetSlice(obj, "spec", "env")
+		if err != nil {
+			return err
+		}
+		volumeMounts, _, err := unstruct.GetSlice(obj, "spec", "volumeMounts")
+		if err != nil {
+			return err
+		}
+
+		if containerName == ContainerOpenEBSCSIPluginName {
+			// Set the image of the container.
+			err = unstructured.SetNestedField(obj.Object, p.ObservedOpenEBS.Spec.CstorConfig.CSI.CSINode.Image,
+				"spec", "image")
+			if err != nil {
+				return err
+			}
+			// Set the environmets of the container.
+			err = unstruct.SliceIterator(envs).ForEachUpdate(updateOpenEBSCSIPluginEnv)
+			if err != nil {
+				return err
+			}
+			err = unstruct.SliceIterator(volumeMounts).ForEachUpdate(updateOpenEBSCSIPluginVolumeMount)
+			if err != nil {
+				return err
+			}
+		}
+		err = unstructured.SetNestedSlice(obj.Object, envs, "spec", "env")
+		if err != nil {
+			return err
+		}
+
+		// Append the new extra volume mounts with the existing volume mounts, required for the csi to work.
+		volumeMounts = append(volumeMounts, extraVolumeMounts...)
+		err = unstructured.SetNestedSlice(obj.Object, volumeMounts, "spec", "volumeMounts")
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// Update the containers.
+	err = unstruct.SliceIterator(containers).ForEachUpdate(updateContainer)
+	if err != nil {
+		return err
+	}
+
+	// Set back the value of the containers.
+	err = unstructured.SetNestedSlice(daemonset.Object,
+		containers, "spec", "template", "spec", "containers")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getOSSpecificVolumeMounts returns the volume and volume mounts based on the specific OS distribution/version.
+// This volume and volume mounts are for the specific container i.e openebs-csi-plugin.
+// This function will get the OS Image and the version for the ubuntu distribution and will return the
+// volumes and volume mounts accordngly.
+func (p *Planner) getOSSpecificVolumeMounts() ([]interface{}, []interface{}, error) {
+	volumes := make([]interface{}, 0)
+	volumeMounts := make([]interface{}, 0)
+
+	osImage, err := k8s.GetOSImage()
+	if err != nil {
+		return volumes, volumeMounts, errors.Errorf("Error getting OS Image of a Node, error: %+v", err)
+	}
+
+	ubuntuVersion, err := k8s.GetUbuntuVersion()
+	if err != nil {
+		return volumes, volumeMounts, errors.Errorf("Error getting Ubuntu Version of a Node, error: %+v", err)
+	}
+
+	switch true {
+	case strings.Contains(strings.ToLower(osImage), strings.ToLower(types.OSImageSLES12)):
+		// Create new volumes for suse 12.
+		libCryptoVolume := map[string]interface{}{
```

Comment:
we can have this as a generic function that takes volume and volumeMounts as arg or maybe a structure will be better and  will just return the formed structure.